capsules/src/analog_comparator.rs

```diff
@@ -132,24 +132,25 @@ impl<'a, A: hil::analog_comparator::AnalogComparator<'a>> LegacyDriver for Analo
             3 => self.stop_comparing(channel),
 
             _ => ReturnCode::ENOSUPPORT,
-        }
+        };
+        retcode.into()
     }
 
     /// Provides a callback which can be used to signal the application
     fn subscribe(
         &self,
         subscribe_num: usize,
-        callback: Option<Callback>,
+        mut callback: Callback,
         _app_id: AppId,
-    ) -> ReturnCode {
+    ) -> Result<Callback, (Callback, ErrorCode)> {
         match subscribe_num {
             // Subscribe to all interrupts
             0 => {
-                self.callback.set(callback);
-                ReturnCode::SUCCESS
+                self.callback.swap(Cell::from_mut(&mut callback));
+                Ok(callback)
```

Comment:
Wait why is this not

```rust
let old_callback = self.callback.replace(callback);
Ok(old_callback)
```

Reply:
it should be, phil suggested the same in #2235 . I didn't realize the swapping itself was not important.